### DISCUSSION TITLE
contributing/development/compiling/compiling_for_linuxbsd

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#256] [User-contributed notes]
# contributing/development/compiling/compiling_for_linuxbsd

Requirements: For compiling under Linux or other Unix variants, the following is required: GCC 9+ or Clang 6+., Python 3.6+., SCons 3.1.2+ build system., pkg-config (used to detect the development ...

https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_linuxbsd.html

**Comment by FlooferLand** [comment]
As of now, Pyston's repo description says its no longer being maintained

**Reply by ghost** [reply]
It's correct but it still works. Even though my latest comparisons shows negligible difference https://github.com/godotengine/godot/pull/98898#issuecomment-2490896984. So generally you can just use python latest and have the "same" results

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10335

**Comment by OS-of-S** [comment]
Here is a script that automates the installation of all the packages and tools mentioned on the page, which also solves a number of problems specific to the old Ubuntu 16.04.

https://forum.godotengine.org/t/how-to-install-everything-to-compile-godot-on-linux-ubuntu-16-4/93092

**Comment by v1993** [comment]
> It's still recommended to use GCC for production builds as they can be compiled using link-time optimization, making the resulting binaries smaller and faster.

Is this actually true? If so, it seems like an issue of some kind and should be noted as such, as LLVM toolchain supports both full and thin LTO just fine - in fact, it's GCC that lacks support for thin LTO.

**Reply by Calinou** [reply]
GCC still performs better optimizations last time we checked (especially with LTO), which is why it's still the recommended compiler for production.

In general, GCC will perform better on some programs and Clang will perform better in others. This is the recurring theme when you look at Phoronix benchmarks. In other words, there isn't a compiler that always performs better than the other.

**Comment by berarma** [comment]
How is the Fedora one-liner installing package embree3-devel when the page says Godot requires Embree 4?

**Reply by Calinou** [reply]
This was fixed in https://github.com/godotengine/godot-docs/pull/10439 but it hasn't been cherry-picked to the `4.3` branch yet.

Install `embree-devel` instead of `embree3-devel`.

**Comment by Vulwsztyn** [comment]
Any change to the cpp code results in need to rebuild that takes about 10 mins? Is there any way to make it bearable?

**Reply by Calinou** [reply]
Rebuilds are automatically incremental, but if you modify a header file (`.h`/`.hpp`), all depending files also have to be recompiled. If you modify `node.h`, `object.h`, `editor_node.h` or any other header that is heavily referenced in other files, expect a full rebuild to occur.

You can set up [ccache](https://ccache.dev/) to speed up incremental rebuilds. Make sure to symlink your `gcc`/`g++`/`clang`/`clang++` to ccache so that SCons uses it automtaically, or use the new SCons options added by https://github.com/godotengine/godot/pull/105498.

**Comment by geekley** [comment]
> Linux binaries usually won't run on distributions that are older than the distribution they were built on. If you wish to distribute binaries that work on most distributions, you should build them on an old distribution such as Ubuntu 16.04. You can use a virtual machine or a container to set up a suitable build environment.

Instead of pointing users to a more complex setup of virtual machines or containers, isn't it better to (also) point them to https://github.com/godotengine/buildroot repo? Assuming that is enough to compile to portable Linux from any Linux. Is it?

**Can someone confirm that by using this, no docker or VM is necessary, e.g. when compiling from the most recent Ubuntu?**

**Comment by nubels** [comment]
Is there a way to cross-compile the export templates for Linux from a macOS device with Apple Silicon? I need this for PCK encryption.

**Reply by Calinou** [reply]
It should be possible to do this directly without needing any virtual machines or containers, but in the worst case scenario, you can use a [Docker container](https://docs.docker.com/desktop/setup/install/mac-install/) with a x86_64 version of Ubuntu 22.04 (`docker run -it ubuntu:22.04 --platform linux/amd64`), clone the Godot source code within and compile it. (Rosetta 2 needs to be installed on your machine for this to work.)

**Reply by nubels** [reply]
Is there a guide somewhere about how to set it up without virtual machines or containers? I'm not sure how to install the other packages.

```
nubels@mbp godot % clang --version                                            
Apple clang version 17.0.0 (clang-1700.0.13.3)
Target: arm64-apple-darwin24.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

nubels@mbp godot % which pkg-config       
/opt/homebrew/bin/pkg-config

nubels@mbp godot % which python3
/opt/homebrew/bin/python3

nubels@mbp godot % which scons
/opt/homebrew/bin/scons

nubels@mbp godot % scons platform=linuxbsd target=template_release arch=x86_64
scons: Reading SConscript files ...
ERROR: Invalid target platform "linuxbsd".
The following platforms are available:
	ios
	macos
	windows
Please run SCons again and select a valid platform: platform=<string>.
```
